### DISCUSSION TITLE
No logs in Events tab

### DISCUSSION THREAD

**Discussion by joinsukesh** [datalust/seq-tickets#1805] [General]
Hello Team,

I've just started with Seq. This is a Sitecore application and uses log4Net as default.

[This is my local machine. I'm using Sitecore 10.2 XP]

This is the patch config, where the appender is declared.

````
<?xml version="1.0" encoding="utf-8" ?><configuration xmlns:patch="http://www.sitecore.net/xmlconfig/" xmlns:set="http://www.sitecore.net/xmlconfig/set/"> 
  <sitecore>  
<log4net>
    <appender name="SeqAppender" type="Seq.Client.Log4Net.SeqAppender, Seq.Client.Log4Net" >
      <serverUrl value="http://localhost:5341" />
      <apiKey value="L7U2ls7FmIxDGwgZIOud" /> 
    </appender>
    <appender name="SeqAsyncForwarder" type="Log4Net.Async.AsyncForwardingAppender,Log4Net.Async">
      <appender-ref ref="SeqAppender" />
    </appender>
    <root>
	  <appender-ref ref="SeqAsyncForwarder" patch:after="*[@ref='LogFileAppender']" />
      <level value="INFO" />
      <!-- <appender-ref ref="SeqAsyncForwarder" /> -->
    </root>
  </log4net>
  </sitecore>
</configuration>
````
The default log files are created but I do not see any entries in Seq.
What am I doing wrong here.

![image](https://user-images.githubusercontent.com/24619393/227950888-853bb49b-e63c-4b92-b6e4-ec6fe16982dd.png)

Thank you

**Comment by liammclennan** [comment]
Hi @sukesh-y 

You are filtering Seq to errors after March 27th. Are you expecting data that matches that restriction?

Have you installed the required nuget packages (Seq.Client.Log4Net, Log4Net.Async)?

Check *Settings > Diagnostics > Ingestion Log* and look for any ingestion errors. 

If you are still having problems try the [log4net internal debugging instructions](https://logging.apache.org/log4net/release/faq.html#internalDebug) and see if you can get an error message.

**Reply by joinsukesh** [reply]
Hello @liammclennan 

Thank you for your reply.
Yes I do have the data as the regular log files from Sitecore are being created.
Yes, I have installed the packages and copied the DLLs to the website /bin folder.

I tried the 2nd approach from the debugging instructions and this is what I see in the log.

```
log4net:ERROR DOMConfigurator: Could not create Appender [SeqAppender] of type [Seq.Client.Log4Net.SeqAppender, Seq.Client.Log4Net]. Reported error follows.
System.InvalidCastException: Unable to cast object of type 'Seq.Client.Log4Net.SeqAppender' to type 'log4net.Appender.IAppender'.
   at log4net.Repository.Hierarchy.DOMHierarchyConfigurator.ParseAppender(XmlElement appenderElement)
```

I came across this answer from Joao Neto, which seems to be related to my issue. 
https://sitecore.stackexchange.com/questions/11196/could-not-create-appender-adonetappender-of-type-log4net-appender-adonetappen
Tried his 2nd approach, which is adding the custom config path in `appSettings`, as I did not know how or what code to copy.

Can **Seq** not read from the already created .txt log files (which are not structured)?
If not, if I create structured .json log files, will Seq be able to read from them?

Sorry, but all this is so frustrating. All I wanted was to a simple log viewer in my local machine to read from the logs :(

**Reply by liammclennan** [reply]
That looks like progress. :)

I also found this [question and answer](https://sitecore.stackexchange.com/questions/17199/use-seq-for-logging-with-log4net-instead-of-text-file) that might be helpful. 

Usually, the interface to Seq is its HTTP API, it doesn't read log files. It is possible to [forward log files to Seq](https://blog.datalust.co/tailing-a-log-file-or-folder-with-fluent-bit-and-seq/), but that is not going to be easier to setup. 

I know how frustrating it can be to get these things working together. It is hard for us to provide support for Sitecore and log4net because we don't control those products.

**Comment by joinsukesh** [comment]
I was able to achieve what I wanted using a different approach.
For some reason, Sitecore was trying to do some parsing with the Seq default DLL.

This is what I have tried.
1. Created a custom class - [SeqLogger.cs](https://github.com/sukesh-y/Basiscore/blob/master/Basiscore.SeqLogger/Services/SeqLogger.cs)

The logic here is, after Sitecore creates the default *.txt log file, the `Append()` method in this class will be triggered.
And in that method, I'm posting the log to Seq's API.

2. And, this is the patch.config which declares my custom class.

```
<?xml version="1.0" encoding="utf-8" ?>
<configuration xmlns:patch="http://www.sitecore.net/xmlconfig/" xmlns:role="http://www.sitecore.net/xmlconfig/role/" xmlns:security="http://www.sitecore.net/xmlconfig/security/">
  <sitecore>    
    <log4net>
      <appender name="SeqLogger" type="Basiscore.SeqLogger.Services.SeqLogger, Basiscore.SeqLogger" patch:after = "appender[@name='LogFileAppender']">
        <applicationname value="Stratum" />
        <includesystemlogs value="false" />
        <!--- excludeloggers: If 'includesystemlogs is 'false' then the logger names starting with any of the specified terms will be excluded from logging-->
        <excludeloggers value="Sitecore.,object,System.RuntimeType"/>
        <seqapiurl value="http://localhost:5341/api/events/raw?clef" />
        <layout type="log4net.Layout.PatternLayout" />
        <encoding value="utf-8" />
      </appender>
      <root>
	    <level value="DEBUG" />
        <appender-ref ref="SeqLogger" patch:after = "*[@ref='LogFileAppender']"/>
      </root>
    </log4net>
  </sitecore>
</configuration>
```

I'm able to view the logs now :)

![image](https://user-images.githubusercontent.com/24619393/228793241-002137c5-d02f-4a63-a284-2a35a7933eee.png)


Thank you for your support

**Comment by liammclennan** [comment]
Hi @sukesh-y 

That's great, and thanks for letting us know.

**Comment by todthomson** [comment]
@sukesh-y for completeness (in case someone else stumbles on to this thread) would you mind posting the following:
1. the Sitecore version (e.g. 10.3), and
2. the Sitecore product that your above code is working with (e.g. Sitecore Experience Manager (XM) or Sitecore Experience Commerce (XC), etc).

Apologies if my question doesn't 100% makes sense, as I've never used Sitecore before.

The reason I bring this up is that Sitecore already provides documentation on how to use Serilog (and therefore Seq, via the Seq sink) to make this work without needing to write custom code.

Commerce Engine configuration file (via `config.json`):
https://doc.sitecore.com/xp/en/developers/103/sitecore-experience-commerce/commerce-engine-configuration-file.html

Logging in Sitecore Host / Sitecore Experience Manager (via `sitecorehost.xml`):
https://doc.sitecore.com/xp/en/developers/103/sitecore-experience-manager/logging.html

Then one would just use the `Serilog.Sinks.Seq` Serilog sink to sink from Serilog to Seq, setting the `serverUrl` and `apiKey` (if that is required) either via `.xml` or `.json` configuration and it all should work automatically:
https://github.com/datalust/serilog-sinks-seq

**Reply by joinsukesh** [reply]
Hello @todthomson 

Thank you for the feedback. 
I'm using Sitecore 10.2 XP and this is my local machine. Updated my question as advised.

Thank you for sharing the URLs, but as I'm using Sitecore XP, they do not apply unfortunately.

About Serilog - I started with the Serilog Nugets initially but experienced 2 issues - [https://stackoverflow.com/questions/75373537/how-to-configure-seq-to-include-exception-stack-trace](https://stackoverflow.com/questions/75373537/how-to-configure-seq-to-include-exception-stack-trace)
[https://sitecore.stackexchange.com/questions/34009/how-to-view-exceptions-in-seq](https://sitecore.stackexchange.com/questions/34009/how-to-view-exceptions-in-seq)


Now, I may be able to fix the "Message/Exception" issue with Serilog, but I have to add my custom class again. 
I want to make my "Sitecore Seq Logger" a reusable module. So, now, I can just share my "Basiscore.SeqLogger.dll & the config file.
Another developer can just copy and it would work.
But, if I use Serilog, then I need to share the Serilog DLLs also.

I hope you got what I'm trying to say.
I guess, I may have to use the Serilog sink someday, but i'll leave that for now :)

**Reply by todthomson** [reply]
Hi @sukesh-y,

Thanks for clarifying your ticket and the product/version of Sitecore you're using.

From looking here, it looks like Sitecore XP 10.3 (it might apply to 10.2 as well) is using Serilog under the hood (though it isn't well documented).

https://doc.sitecore.com/xp/en/developers/103/sitecore-experience-platform/configure-the-log-file-size.html

I have to proclaim ignorance here, as I'm not sure what XML format Sitecore are using for their Serilog config. The two I'm familiar with are the [XML-based `<appSettings>` configuration](https://github.com/datalust/serilog-sinks-seq#xml-appsettings-configuration) and [JSON-based `appsettings.json` configuration](https://github.com/datalust/serilog-sinks-seq#json-appsettingsjson-configuration).

That being said, you might like to try having a go at configuring `sc.Serilog.xml` to write to Seq directly, by adding in something like (completely guessing at the syntax here):

```XML
<WriteTo>
    <Seq>
        <Args>
            <serverUrl>http://localhost:5341</serverUrl>
            <apiKey>[optional API key here]</apiKey>
        </Args>
    </Seq>
</WriteTo> 
```

The main challenge here is how to configure Sitecore XP's Serilog config, rather than getting Serilog to talk to Seq. So my suggestion would be to raise a support ticket with Sitecore XP, link them to their Sitecore XP documentation page (above), and ask them how to define a custom Serilog sinks in Sitecore XP configuration file `sc.Serilog.xml`.

As you can see [here](https://github.com/serilog/serilog/wiki/Provided-Sinks), the are a bucket-load of Serilog sinks, and they are all configured in Serilog in much the same way, so I think some good examples from Sitecore on how to add extra Serilog sinks to Sitecore XP would get you a long way to a better solution than going via `log4net` being that Serilog is a replacement for log4net.

Cheers,

Tod.

**Comment by joinsukesh** [comment]
@liammclennan @todthomson 
On a totally different note, is the Seq team/group available on Stackoverlow & Slack.
If not, then kindly consider.

**Reply by todthomson** [reply]
Hi @sukesh-y,

Thanks for the suggestions! :)

There's currently no public Slack or Discord (etc) yet, but I will discuss the idea of one internally with the Seq team.

Unfortunately, the `Seq` tag on Stack Overflow (SO) is currently an alias to `Sequence`. I have requested they remove the alias so SO users can tag Seq tickets with `Seq` and we'll see them when they come in. Currently, we do often track the `Serilog` tag on SO looking for anyone having challenges with Seq, and we help if we can.

As a general statement, the best way to get support is to have an [active paid Seq subscription](https://datalust.co/pricing), and then:
- post here (as you have done), as this is the official place to get support on the product, or
- additionally, paid subscribers have access to an alternative support process where they can raise requests with us directly.

Cheers,

Tod.
